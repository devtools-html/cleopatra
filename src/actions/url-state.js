/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
