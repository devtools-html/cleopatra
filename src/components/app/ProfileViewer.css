@keyframes profileViewerFadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

.profileViewer {
  animation-duration: 200ms;
  animation-name: profileViewerFadeIn;
  background-color: #fff;
}

.profileViewerName {
  height: 24px;
  padding: 0 6px;
  border-right: 1px solid var(--grey-30);
  color: var(--grey-60);
  font-weight: bold;
  line-height: 24px;
}

.profileViewerZipButton {
  /* Position */
  width: 18px;
  height: 18px;

  /* Box */
  flex: none;
  padding: 0;
  margin: 3px 0 3px 3px;
  border: 1px solid var(--green-60);

  /* Other */
  background: var(--green-50) url(../../../res/img/svg/back-arrow.svg) center
    center no-repeat;
  border-radius: 3px;
  color: #000;
}

.profileViewerZipButton:hover {
  border-color: #000;
}

.profileViewerSplitter {
  /* Reset the position: absolute that is set by the splitter-layout class */
  position: unset;
}

.profileViewerSplitter .layout-pane:not(.layout-pane-primary) {
  display: flex;
  overflow: hidden;
  max-height: var(--profile-viewer-splitter-max-height);
  flex-direction: column;
}
