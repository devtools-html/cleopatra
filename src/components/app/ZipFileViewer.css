/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.zipFileViewer {
  width: 100%;
  display: flex;
  justify-content: center;
  min-height: 100vh;
  align-items: center;
  line-height: 1.5;
}

.zipFileViewerHeader {
  font-size: 130%;
  padding: 0 20px;
}

.zipFileViewerSection {
  position: relative;
  width: 75vw;
  height: 75vh;
  max-width: 1200px;

  display: flex;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 3px;

  flex-direction: column;
  box-shadow: 0 5px 25px #0b1f50;
  background-color: #fff;
  overflow: hidden;
}

.zipFileViewerSection > .treeView {
  font-size: 14px;
}

.zipFileViewerSection .treeRowToggleButton {
  margin: 0 5px;
}

.zipFileViewerMessage {
  display: flex;
  flex: 1;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  border-top: 1px solid var(--grey-30);
  background-color: #f5f5f5;
  font-size: 14px;
}

.zipFileViewerMessage > * {
  margin: 10px 30px;
}

.zipFileViewerUntrustedFilePath {
  /**
   * Make the max width 50% so that some random user can't spam a bunch of misleading
   * content into an error message on perf.html.
   */
  max-width: 70%;
  background-color: var(--grey-30);
  padding: 0 5px;
  display: inline-block;
  font-family: monospace;
  overflow: hidden;
  text-overflow: ellipsis;
  border-radius: 3px;
}
