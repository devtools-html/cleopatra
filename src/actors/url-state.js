/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
