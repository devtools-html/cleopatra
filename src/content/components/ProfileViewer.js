import React, { PureComponent, PropTypes } from 'react';
import { connect } from 'react-redux';
import TabBar from '../components/TabBar';
import ProfileSummaryView from '../containers/ProfileSummaryView';
import ProfileCallTreeView from '../containers/ProfileCallTreeView';
import ProfileMarkersView from '../containers/ProfileMarkersView';
import ProfileTaskTracerView from '../containers/ProfileTaskTracerView';
import ProfileLogView from '../containers/ProfileLogView';
import ProfileFilterNavigator from '../containers/ProfileFilterNavigator';
import ProfileSharing from '../containers/ProfileSharing';
import SymbolicationStatusOverlay from '../containers/SymbolicationStatusOverlay';
import TimelineView from '../containers/TimelineView';
import actions from '../actions';
import { getProfileViewOptions, getDisplayRange } from '../reducers/profile-view';
import { getSelectedTab } from '../reducers/url-state';
import ProfileViewerHeader from '../containers/ProfileViewerHeader';

class ProfileViewer extends PureComponent {
  constructor(props) {
    super(props);
    this._onSelectTab = this._onSelectTab.bind(this);

    // If updating this list, make sure and update the tabOrder reducer with another index.
    this._tabs = [
      {
        name: 'summary',
        title: 'Summary',
      },
      {
        name: 'calltree',
        title: 'Call Tree',
      },
      {
        name: 'markers',
        title: 'Markers',
      },
      {
        name: 'tasktracer',
        title: 'Task Tracer',
      },
      {
        name: 'log',
        title: 'Log',
      },
      {
        name: 'timeline',
        title: 'Timeline',
      },
    ];
  }

  _onSelectTab(selectedTab) {
    const { changeSelectedTab } = this.props;
    changeSelectedTab(selectedTab);
  }

  render() {
    const {
      className, viewOptions, timeRange, changeTabOrder, selectedTab,
    } = this.props;
    const { tabOrder } = viewOptions;
    return (
      <div className={className}>
        <div className={`${className}TopBar`}>
          <ProfileFilterNavigator />
          <ProfileSharing />
        </div>
        <ProfileViewerHeader />
        <TabBar tabs={this._tabs}
                selectedTabName={selectedTab}
                tabOrder={tabOrder}
                onSelectTab={this._onSelectTab}
                onChangeTabOrder={changeTabOrder} />
        {{
          summary: <ProfileSummaryView />,
          calltree: <ProfileCallTreeView />,
          markers: <ProfileMarkersView />,
          tasktracer: <ProfileTaskTracerView rangeStart={timeRange.start} rangeEnd={timeRange.end} />,
          timeline: <TimelineView />,
          log: <ProfileLogView />,
        }[selectedTab]}
        <SymbolicationStatusOverlay />
      </div>
    );
  }
}

ProfileViewer.propTypes = {
  className: PropTypes.string.isRequired,
  viewOptions: PropTypes.object.isRequired,
  timeRange: PropTypes.object.isRequired,
  selectedTab: PropTypes.string.isRequired,
  changeSelectedTab: PropTypes.func.isRequired,
  changeTabOrder: PropTypes.func.isRequired,
};

export default connect(state => ({
  viewOptions: getProfileViewOptions(state),
  selectedTab: getSelectedTab(state),
  className: 'profileViewer',
  timeRange: getDisplayRange(state),
}), actions)(ProfileViewer);
