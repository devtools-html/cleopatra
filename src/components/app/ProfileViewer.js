/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import explicitConnect from '../../utils/connect';
import DetailsContainer from './DetailsContainer';
import ProfileFilterNavigator from './ProfileFilterNavigator';
import MenuButtons from './MenuButtons';
import SymbolicationStatusOverlay from './SymbolicationStatusOverlay';
import { returnToZipFileList } from '../../actions/zipped-profiles';
import { getProfileName } from '../../selectors/url-state';
import Timeline from '../timeline';
import { getHasZipFile } from '../../selectors/zipped-profiles';
import SplitterLayout from 'react-splitter-layout';
import { invalidatePanelLayout } from '../../actions/app';
import { getTimelineHeight } from '../../selectors/app';

import type { CssPixels } from '../../types/units';
import type { ConnectedProps } from '../../utils/connect';

require('./ProfileViewer.css');

type StateProps = {|
  +profileName: string | null,
  +hasZipFile: boolean,
  +timelineHeight: CssPixels | null,
|};

type DispatchProps = {|
  +returnToZipFileList: typeof returnToZipFileList,
  +invalidatePanelLayout: typeof invalidatePanelLayout,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class ProfileViewer extends PureComponent<Props> {
  render() {
    const {
      hasZipFile,
      profileName,
      returnToZipFileList,
      invalidatePanelLayout,
      timelineHeight,
    } = this.props;

    return (
      <div
        className="profileViewer"
        style={
          timelineHeight === null
            ? {}
            : {
                '--profile-viewer-splitter-max-height': `${timelineHeight}px`,
              }
        }
      >
        <div className="profileViewerTopBar">
          {hasZipFile ? (
            <button
              type="button"
              className="profileViewerZipButton"
              title="View all files in the zip file"
              onClick={returnToZipFileList}
            />
          ) : null}
          {profileName ? (
            <div className="profileViewerName">{profileName}</div>
          ) : null}
          <ProfileFilterNavigator />
          {
            // Define a spacer in the middle that will shrink based on the availability
            // of space in the top bar. It will shrink away before any of the items
            // with actual content in them do.
          }
          <div className="profileViewerSpacer" />
          <MenuButtons />
        </div>
        <SplitterLayout
          customClassName="profileViewerSplitter"
          vertical
          percentage={false}
          // The DetailsContainer is primary.
          primaryIndex={1}
          // The Timeline is secondary.
          secondaryInitialSize={270}
          onDragEnd={invalidatePanelLayout}
        >
          <Timeline />
          <DetailsContainer />
        </SplitterLayout>
        <SymbolicationStatusOverlay />
      </div>
    );
  }
}

export default explicitConnect<{||}, StateProps, DispatchProps>({
  mapStateToProps: state => ({
    profileName: getProfileName(state),
    hasZipFile: getHasZipFile(state),
    timelineHeight: getTimelineHeight(state),
  }),
  mapDispatchToProps: {
    returnToZipFileList,
    invalidatePanelLayout,
  },
  component: ProfileViewer,
});
