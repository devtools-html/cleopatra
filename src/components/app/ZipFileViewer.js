/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import explicitConnect, {
  type ExplicitConnectOptions,
  type ConnectedProps,
} from '../../utils/connect';
import {
  procureInitialInterestingExpandedNodes,
  type ZipFileTree,
  type IndexIntoZipFileTable,
} from '../../profile-logic/zip-files';
import {
  changeSelectedZipFile,
  changeExpandedZipFile,
  viewProfileFromZip,
  returnToZipFileList,
  showErrorForNoFileInZip,
} from '../../actions/app';
import {
  getZipFileState,
  getZipFileTree,
  getZipFileTable,
  getZipFileMaxDepth,
  getSelectedZipFileIndex,
  getExpandedZipFileIndexes,
} from '../../reducers/app';
import { getZipFilePathFromUrl } from '../../reducers/url-state';
import TreeView from '../shared/TreeView';
import ProfileViewer from './ProfileViewer';
import type { ZipFileState } from '../../types/reducers';
import type {
  ZipFileTable,
  ZipDisplayData,
} from '../../profile-logic/zip-files';
import './ZipFileViewer.css';

type StateProps = {|
  +zipFileState: ZipFileState,
  +zipFilePathFromUrl: string | null,
  +zipFileTree: ZipFileTree,
  +zipFileTable: ZipFileTable,
  +zipFileMaxDepth: number,
  +selectedZipFileIndex: IndexIntoZipFileTable | null,
  // In practice this should never contain null, but needs to support the
  // TreeView interface.
  +expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>,
|};

type DispatchProps = {|
  +changeSelectedZipFile: typeof changeSelectedZipFile,
  +changeExpandedZipFile: typeof changeExpandedZipFile,
  +viewProfileFromZip: typeof viewProfileFromZip,
  +returnToZipFileList: typeof returnToZipFileList,
  +showErrorForNoFileInZip: typeof showErrorForNoFileInZip,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

type ZipFileSharedRowProps = {|
  +viewProfileFromZip: typeof viewProfileFromZip,
|};
type ZipFileRowProps = {|
  +sharedRowProps: ZipFileSharedRowProps,
  +displayData: ZipDisplayData,
|};

class ZipFileRow extends React.PureComponent<ZipFileRowProps> {
  _handleClick = (event: SyntheticMouseEvent<HTMLElement>) => {
    if (event.metaKey || event.ctrlKey) {
      return;
    }
    event.preventDefault();
    const {
      sharedRowProps: { viewProfileFromZip },
      displayData: { zipTableIndex },
    } = this.props;
    if (zipTableIndex !== null) {
      viewProfileFromZip(zipTableIndex);
    }
  };
  render() {
    const { name, url } = this.props.displayData;
    if (!url) {
      return name;
    }

    return (
      <a href={url} onClick={this._handleClick}>
        {name}
      </a>
    );
  }
}

/**
 * This component is a viewer for zip files. It was built to load
 * multiple profiles recorded from Talos so that a user could navigate
 * to a particular one. However, it is a general purpose zip file
 * viewer to load profiles, so it can be used on arbitrary profiles.
 */
class ZipFileViewer extends React.PureComponent<Props> {
  _fixedColumns = [];
  _mainColumn = { propName: 'name', title: '', component: ZipFileRow };
  _appendageButtons = ['focusCallstackButton'];
  _treeView: ?TreeView<
    IndexIntoZipFileTable,
    ZipDisplayData,
    ZipFileSharedRowProps
  >;
  _takeTreeViewRef = treeView => (this._treeView = treeView);

  componentWillMount() {
    const {
      expandedZipFileIndexes,
      zipFileTree,
      changeExpandedZipFile,
    } = this.props;
    if (expandedZipFileIndexes.length === 0 && zipFileTree) {
      changeExpandedZipFile(
        procureInitialInterestingExpandedNodes(zipFileTree)
      );
    }
  }

  componentDidMount() {
    const { zipFileState, zipFilePathFromUrl, zipFileTree } = this.props;
    if (zipFileState.phase === 'NONE' && zipFilePathFromUrl) {
      // Most likely the UrlState was deserialized from the URL, but the zip file
      // still hasn't actually been decompressed yet.
      if (!zipFileTree) {
        throw new Error(
          'The zipFileTree should exist if this component was mounted'
        );
      }
    }
    this.focus();
  }

  /**
   * This function keeps the ZipFileState in sync with the UrlState.
   * Loading zip files is asynchronous, while navigating URLs can
   * happen faster than zip files can be read and processed. This
   * method is what keeps the ZipFileViewer and ZipFileState in sync
   * with the UrlState.
   */
  componentWillReceiveProps(nextProps: Props) {
    const {
      zipFilePathFromUrl,
      zipFileState,
      viewProfileFromZip,
      zipFileTable,
      returnToZipFileList,
      showErrorForNoFileInZip,
    } = nextProps;

    if (zipFilePathFromUrl !== zipFileState.zipFilePath) {
      // The UrlState and ZipFileState are out of sync, they need to be
      // updated.

      if (zipFilePathFromUrl === null) {
        // The the UrlState has navigated away from the file, return
        // to the zip file list.
        returnToZipFileList();
      } else {
        // The UrlState was updated to view a zip file, but we haven't
        // started doing that.
        const zipFileIndex = zipFileTable.path.indexOf(zipFilePathFromUrl);
        if (zipFileIndex === -1) {
          showErrorForNoFileInZip(zipFilePathFromUrl);
        } else {
          viewProfileFromZip(zipFileIndex);
        }
      }
    }
  }

  focus() {
    const treeView = this._treeView;
    if (treeView) {
      treeView.focus();
    }
  }

  _onAppendageButtonClick = (zipFileIndex: IndexIntoZipFileTable | null) => {
    if (zipFileIndex !== null) {
      this.props.viewProfileFromZip(zipFileIndex);
    }
  };

  _renderMessage(message: React.Node) {
    return (
      <section className="zipFileViewer">
        <div className="zipFileViewerSection">
          <header className="zipFileViewerHeader">
            <h1>perf.html</h1>
            <p>Choose a profile from this zip file</p>
          </header>
          <div className="zipFileViewerMessage">{message}</div>
        </div>
      </section>
    );
  }

  _renderBackButton() {
    return (
      <button
        key="backButton"
        type="button"
        className="homeSectionInstallButton"
        onClick={this.props.returnToZipFileList}
      >
        ‚Üê Back to zip file list
      </button>
    );
  }

  render() {
    const {
      zipFileState,
      zipFileTree,
      zipFileMaxDepth,
      selectedZipFileIndex,
      expandedZipFileIndexes,
      changeSelectedZipFile,
      changeExpandedZipFile,
      zipFilePathFromUrl,
      viewProfileFromZip,
    } = this.props;

    if (!zipFileTree) {
      console.error('No zipFileTree was found in a ZipFileViewer.');
      return null;
    }
    const { phase } = zipFileState;

    switch (phase) {
      case 'NO_ZIP_FILE':
        console.error(
          'Loaded the ZipFileViewer component when there is no zip file.'
        );
        return this._renderMessage(<span>Error: No zip file was found.</span>);
      case 'LIST_FILES_IN_ZIP_FILE':
        return (
          <section className="zipFileViewer">
            <div className="zipFileViewerSection">
              <header className="zipFileViewerHeader">
                <h1>perf.html</h1>
                <p>Choose a profile from this zip file</p>
              </header>
              <TreeView
                maxNodeDepth={zipFileMaxDepth}
                tree={zipFileTree}
                fixedColumns={this._fixedColumns}
                mainColumn={this._mainColumn}
                onSelectionChange={changeSelectedZipFile}
                onExpandedNodesChange={changeExpandedZipFile}
                selectedNodeId={selectedZipFileIndex}
                expandedNodeIds={expandedZipFileIndexes}
                appendageButtons={this._appendageButtons}
                onAppendageButtonClick={this._onAppendageButtonClick}
                ref={this._takeTreeViewRef}
                contextMenuId={'MarkersContextMenu'}
                rowHeight={30}
                indentWidth={15}
                sharedRowProps={{
                  viewProfileFromZip,
                }}
              />
            </div>
          </section>
        );
      case 'PROCESS_PROFILE_FROM_ZIP_FILE':
        return this._renderMessage(
          <span>Loading the profile from the zip file...</span>
        );
      case 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE':
        return this._renderMessage([
          <span key="message">
            Are you sure this is a profile? Failed to process the file in the
            zip file at the following path:
          </span>,
          <span className="zipFileViewerUntrustedFilePath" key="path">
            /{
              // This is text that comes from the URL, make sure it looks visually
              // distinct for guarding against someone doing anything too funky
              // with it.
              zipFilePathFromUrl
            }
          </span>,
          this._renderBackButton(),
        ]);
      case 'FILE_NOT_FOUND_IN_ZIP_FILE':
        return this._renderMessage([
          <span key="message">
            Failed to find a file in the zip at the following path:
          </span>,
          <span className="zipFileViewerUntrustedFilePath" key="path">
            /{
              // This is text that comes from the URL, make sure it looks visually
              // distinct for guarding against someone doing anything too funky
              // with it.
              zipFilePathFromUrl
            }
          </span>,
          this._renderBackButton(),
        ]);
      case 'VIEW_PROFILE_IN_ZIP_FILE':
        return <ProfileViewer />;
      default:
        (phase: empty); // eslint-disable-line no-unused-expressions
        throw new Error('Unknown zip file phase.');
    }
  }
}

const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
  mapStateToProps: state => {
    const zipFileTree = getZipFileTree(state);
    if (zipFileTree === null) {
      throw new Error(
        'The zipFileTree should exist if the ZipFileViewer is mounted.'
      );
    }
    const zipFileTable = getZipFileTable(state);
    if (zipFileTable === null) {
      throw new Error(
        'The zipFileTable should exist if the ZipFileViewer is mounted.'
      );
    }
    return {
      zipFileState: getZipFileState(state),
      zipFilePathFromUrl: getZipFilePathFromUrl(state),
      zipFileTable: zipFileTable,
      zipFileTree,
      zipFileMaxDepth: getZipFileMaxDepth(state),
      selectedZipFileIndex: getSelectedZipFileIndex(state),
      expandedZipFileIndexes: getExpandedZipFileIndexes(state),
    };
  },
  mapDispatchToProps: {
    changeSelectedZipFile,
    changeExpandedZipFile,
    viewProfileFromZip,
    returnToZipFileList,
    showErrorForNoFileInZip,
  },
  component: ZipFileViewer,
};

export default explicitConnect(options);
