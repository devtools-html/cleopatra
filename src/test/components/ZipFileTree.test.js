/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import ZipFileViewer from '../../components/app/ZipFileViewer';
import ProfileViewer from '../../components/app/ProfileViewer';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { storeWithZipFile } from '../fixtures/profiles/zip-file';
import * as UrlStateSelectors from '../../selectors/url-state';
import * as ZippedProfileSelectors from '../../selectors/zipped-profiles';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { waitUntilState } from '../fixtures/utils';

describe('calltree/ZipFileTree', function() {
  async function setup() {
    const { store } = await storeWithZipFile([
      'foo/bar/profile1.json',
      'foo/profile2.json',
      'baz/profile3.json',
    ]);

    // Some child components render to canvas.
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => mockCanvasContext());

    const component = mount(
      <Provider store={store}>
        <ZipFileViewer />
      </Provider>
    );
    const { getState, dispatch } = store;
    return { store, component, getState, dispatch };
  }
  it('renders a zip file tree', async () => {
    const { component } = await setup();
    expect(component).toMatchSnapshot();
  });

  it('contains a list of all the files', async () => {
    const { component } = await setup();

    const rowNames = component
      .find('.treeViewRow')
      .map(component => component.text())
      .filter(text => text);

    expect(rowNames).toEqual([
      'foo',
      'bar',
      'profile1.json',
      'profile2.json',
      'baz',
      'profile3.json',
    ]);
  });

  describe('clicking on a profile link', function() {
    const setupClickingTest = async () => {
      const { component, store, getState, dispatch } = await setup();

      const profile1OpenLink = component
        .find('a')
        .filterWhere(component => component.text() === 'profile1.json');

      const waitUntilDoneProcessingZip = () =>
        waitUntilState(
          store,
          state =>
            ZippedProfileSelectors.getZipFileState(state).phase ===
            'VIEW_PROFILE_IN_ZIP_FILE'
        );

      return {
        component,
        store,
        getState,
        dispatch,
        profile1OpenLink,
        waitUntilDoneProcessingZip,
      };
    };

    it('starts out without any path in the zip file', async () => {
      const { getState, component } = await setupClickingTest();
      expect(UrlStateSelectors.getPathInZipFileFromUrl(getState())).toBe(null);
      expect(component.find(ProfileViewer).length).toBe(0);
    });

    it('kicks off the profile loading process when clicked', async () => {
      const {
        component,
        getState,
        profile1OpenLink,
        waitUntilDoneProcessingZip,
      } = await setupClickingTest();
      profile1OpenLink.simulate('click');
      component.update();
      expect(UrlStateSelectors.getPathInZipFileFromUrl(getState())).toBe(
        'foo/bar/profile1.json'
      );

      // The profile isn't actually viewed yet.
      expect(component.find(ProfileViewer).length).toBe(0);
      expect(component.find('.zipFileViewerMessage').text()).toMatchSnapshot();

      // There is async behavior going on. Wait until it is done or else
      // redux-react will throw an error.
      await waitUntilDoneProcessingZip();
    });

    it('will show the ProfileViewer component when done processing', async () => {
      // The full ProfileViewer component renders to a Canvas, which the jsdom API
      // doesn't fully mock out.
      jest
        .spyOn(HTMLCanvasElement.prototype, 'getContext')
        .mockImplementation(() => {
          return mockCanvasContext();
        });

      const {
        component,
        profile1OpenLink,
        waitUntilDoneProcessingZip,
      } = await setupClickingTest();

      profile1OpenLink.simulate('click');
      await waitUntilDoneProcessingZip();

      component.update();
      expect(component.find(ProfileViewer).length).toBe(1);
    });
  });
});
