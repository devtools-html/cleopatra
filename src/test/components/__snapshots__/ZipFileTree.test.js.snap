// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calltree/ZipFileTree clicking on a profile link kicks off the profile loading process when clicked 1`] = `"Loading the profile from the zip file..."`;

exports[`calltree/ZipFileTree renders a zip file tree 1`] = `
<section
  class="zipFileViewer"
>
  <div
    class="zipFileViewerSection"
  >
    <header
      class="zipFileViewerHeader"
    >
      <h1>
        Firefox Profiler
      </h1>
      <p>
        Choose a profile from this zip file
      </p>
    </header>
    <div
      class="treeView"
    >
      <div
        class="react-contextmenu-wrapper treeViewContextMenu"
      >
        <div
          aria-label="Call tree"
          class="treeViewBody"
          role="tree"
          tabindex="0"
        >
          <div
            class="treeViewBodyInnerWrapper"
          >
            <div
              class="treeViewBodyInner treeViewBodyInner0"
              style="height: 180px;"
            >
              <div
                class="treeViewBodyInner treeViewBodyInner0TopSpacer"
                style="height: 0px;"
              />
              <div
                class="treeViewBodyInner treeViewBodyInner0InnerChunk"
              >
                <div
                  class="treeViewRow treeViewRowFixedColumns even "
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd "
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even "
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd "
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns even "
                  style="height: 30px; line-height: 30px;"
                />
                <div
                  class="treeViewRow treeViewRowFixedColumns odd "
                  style="height: 30px; line-height: 30px;"
                />
              </div>
            </div>
            <div
              class="treeViewBodyInner treeViewBodyInner1"
              style="height: 180px; width: 3000px;"
            >
              <div
                class="treeViewBodyInner treeViewBodyInner1TopSpacer"
                style="height: 0px;"
              />
              <div
                class="treeViewBodyInner treeViewBodyInner1InnerChunk"
              >
                <div
                  aria-expanded="true"
                  aria-label="foo, running time is undefined (undefined), self time is undefined"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    foo
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-label="bar, running time is undefined (undefined), self time is undefined"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    bar
                  </span>
                </div>
                <div
                  aria-label="profile1.json, running time is undefined (undefined), self time is undefined"
                  aria-level="3"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 30px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=foo%2Fbar%2Fprofile1.json"
                    >
                      profile1.json
                    </a>
                  </span>
                </div>
                <div
                  aria-label="profile2.json, running time is undefined (undefined), self time is undefined"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=foo%2Fprofile2.json"
                    >
                      profile2.json
                    </a>
                  </span>
                </div>
                <div
                  aria-expanded="true"
                  aria-label="baz, running time is undefined (undefined), self time is undefined"
                  aria-level="1"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns even  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 0px;"
                  />
                  <span
                    class="treeRowToggleButton expanded canBeExpanded"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    baz
                  </span>
                </div>
                <div
                  aria-label="profile3.json, running time is undefined (undefined), self time is undefined"
                  aria-level="2"
                  aria-selected="false"
                  class="treeViewRow treeViewRowScrolledColumns odd  "
                  role="treeitem"
                  style="height: 30px; line-height: 30px;"
                >
                  <span
                    class="treeRowIndentSpacer"
                    style="width: 15px;"
                  />
                  <span
                    class="treeRowToggleButton collapsed leaf"
                  />
                  <span
                    class="treeViewRowColumn treeViewMainColumn name"
                  >
                    <a
                      href="http://localhost/from-url//calltree/?file=baz%2Fprofile3.json"
                    >
                      profile3.json
                    </a>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
